app/upgrades/v18/upgrades.go

```diff
@@ -0,0 +1,216 @@
+		outerMax := hostZone.RedemptionRate.Add(outerMaxDelta)
+
+		hostZone.MinRedemptionRate = outerMin
+		hostZone.MaxRedemptionRate = outerMax
+
+		k.SetHostZone(ctx, hostZone)
+	}
+}
+
+// Modify HostZoneUnbonding and UserRedemptionRecords NativeTokenAmount to reflect new data structs
+func UpdateUnbondingRecords(ctx sdk.Context, k stakeibckeeper.Keeper, rk recordskeeper.Keeper) error {
+	for _, epochUnbondingRecord := range rk.GetAllEpochUnbondingRecord(ctx) {
+		// loop over host zone unbonding records
+		for _, hostZoneUnbonding := range epochUnbondingRecord.GetHostZoneUnbondings() {
+			recordStatus := hostZoneUnbonding.Status
+			if recordStatus != recordtypes.HostZoneUnbonding_EXIT_TRANSFER_QUEUE {
+				continue // can ignore this record, as it's not unbonding
+			}
+			amountUnbonded := sdkmath.ZeroInt()
+			// check if the unbond amount is stored in cache above
+			allAmountsUnbondedForHost, found := TotalAmountUnbonded[hostZoneUnbonding.HostZoneId]
+			if !found {
+				// check if this unbond time is in the cache above
+				unbondTime := strconv.FormatUint(hostZoneUnbonding.UnbondingTime, 10)
+				for unbondTimePrefixInCache, amountUnbondedInCache := range allAmountsUnbondedForHost {
+					if strings.HasPrefix(unbondTime, unbondTimePrefixInCache) {
+						amountUnbonded = amountUnbondedInCache
+						break
+					}
+				}
+			}
+			// if we can't find unbonding amount, then we use an estimated redemption rate
+			if amountUnbonded.IsZero() {
+				initialRedemptionRate := InitialRedemptionRates[hostZoneUnbonding.HostZoneId]
+				hostZone, found := k.GetHostZone(ctx, hostZoneUnbonding.HostZoneId)
+				if !found {
+					return errorsmod.Wrapf(stakeibctypes.ErrHostZoneNotFound, "unable to find host zone with chain-id %s", hostZoneUnbonding.HostZoneId)
+				}
+				currentRedemptionRate := hostZone.RedemptionRate
+				blendedRedemptionRate := initialRedemptionRate.Add(currentRedemptionRate).Quo(sdk.NewDecFromInt(sdk.NewInt(2)))
+				amountUnbonded = sdk.NewDecFromInt(hostZoneUnbonding.StTokenAmount).Mul(blendedRedemptionRate).TruncateInt()
+			}
+			hostZoneUnbonding.NativeTokenAmount = amountUnbonded
+			// get implied redemption rate, based on the amount unbonded and stTokens burned
+			impliedRedemptionRate := sdk.NewDecFromInt(amountUnbonded).Quo(sdk.NewDecFromInt(hostZoneUnbonding.StTokenAmount))
+			// now update all userRedemptionRecords
+			for _, userRedemptionRecordId := range hostZoneUnbonding.GetUserRedemptionRecords() {
+				userRedemptionRecord, found := rk.GetUserRedemptionRecord(ctx, userRedemptionRecordId)
+				if !found {
+					return errorsmod.Wrapf(recordtypes.ErrHostUnbondingRecordNotFound, "unable to find user redemption record with id %s", userRedemptionRecordId)
+				}
+				userRedemptionRecord.NativeTokenAmount = sdk.NewDecFromInt(userRedemptionRecord.StTokenAmount).Mul(impliedRedemptionRate).TruncateInt()
+				rk.SetUserRedemptionRecord(ctx, userRedemptionRecord)
+			}
```

Comment:
do we have to do a catch all here on the last record to make sure there's no precision error?

Reply:
Ah yes that's a great point, we should update it to that!